src/main/kotlin/hwr/oop/projects/peakpoker/core/hand/HandEvaluator.kt

```diff
@@ -1,47 +1,83 @@
-            isStraight && isFlush -> HandRank.STRAIGHT_FLUSH
-            rankCounts.first() == 4 -> HandRank.FOUR_OF_A_KIND
-            rankCounts.first() == 3 && rankCounts[1] == 2 -> HandRank.FULL_HOUSE
-            isFlush -> HandRank.FLUSH
-            isStraight -> HandRank.STRAIGHT
-            rankCounts.first() == 3 -> HandRank.THREE_OF_A_KIND
-            rankCounts.first() == 2 && rankCounts[1] == 2 -> HandRank.TWO_PAIR
-            rankCounts.first() == 2 -> HandRank.ONE_PAIR
-            else -> HandRank.HIGH_CARD
+            isStraight && isFlush                                          -> HandRank.STRAIGHT_FLUSH
+            rankCounts[0] == 4                                             -> HandRank.FOUR_OF_A_KIND
+            rankCounts[0] == 3 && rankCounts[1] == 2                       -> HandRank.FULL_HOUSE
+            isFlush                                                        -> HandRank.FLUSH
+            isStraight                                                     -> HandRank.STRAIGHT
+            rankCounts[0] == 3                                             -> HandRank.THREE_OF_A_KIND
+            rankCounts[0] == 2 && rankCounts[1] == 2                       -> HandRank.TWO_PAIR
+            rankCounts[0] == 2                                             -> HandRank.ONE_PAIR
+            else                                                           -> HandRank.HIGH_CARD
+        }
+    }
+
+    /**
+     * Combines hole cards and community cards, evaluates all 5-card combinations,
+     * and returns the highest HandRank.
+     */
+    fun evaluateAll(hole: HoleCards, community: CommunityCards): HandRank {
+        val allCards = hole.cards + community.cards
+        require(allCards.size == 7) { "Total cards must be 7 (2 hole + 5 community)" }
+                                                        //Bitmasking method
+        var bestRank = HandRank.HIGH_CARD               //default
+        val n = allCards.size                           //will be 7 anyway
+        val totalMasks = 1 shl n                        // 128 = 2^7 (7 = number of cards) -> possible combinations to choose from and evaluate
+        for (mask in 0 until totalMasks) {
+            if (Integer.bitCount(mask) == 5) {      //whenever we are looking at a combination of 5 cards (when 5 bits are set to 1) ->(7 über 5 = 21 valid combinations)
+                val combo = mutableListOf<Card>()
+                for (i in 0 until n) {
+                    if (mask shr i and 1 == 1) combo += allCards[i] //if the i-th bit is set to 1, add the i-th card to the combination
+                }                                                   // shr looks at the bits of the number and shifts them to the right in this case 1 at a time
+                val rank = evaluate(combo)                   // since we extracted it now, we can use our eval function again easily
```

Comment:
If we still need those two functions, `evaluate()` will take `CommunityCards` as the argument type.
Therefore you would need to parse the `CommunityCards` provided by the `evaluateAll()` arguments.

```suggestion
                val rank = evaluate(CommunityCards(cards=combo.toList(), community.game))                   // since we extracted it now, we can use our eval function again easily
```

Reply:
not sure if i understand this correctly but combo will not only consist of community cards, so id keep it this way, let me know on whatsapp!